### PR TITLE
Bump the PyTorch pin to 2.14

### PR DESCRIPTION
**Pull request by shoumikhin**
## Summary

PyTorch 2.14.0 is released, so move the pin from 2.13 to it. The tree is mid-cycle at 1.5.0 and the pin normally follows the current stable release.

The pin itself is a few version strings plus a re-sync of the vendored c10 headers, which CI requires to match PyTorch's tree byte for byte. Everything else here is a change 2.14 forces, each as its own commit.

## What 2.14 forces

**C++20 where ATen headers are compiled.** `intrusive_ptr.h` uses `operator<=>` and `TensorBase.h` uses `requires`, neither guarded, so a target that includes ATen cannot parse at C++17. This follows the pattern already in the tree: six CMake targets set `CXX_STANDARD 20` for exactly this reason. Added the Vulkan op tests to that list, and taught the Buck wrapper to make the same per-target distinction. Non-ATen targets, including everything the bare-metal presets build, still compile at C++17.

**Two new AOTI entry points.** 2.14's generated wrapper calls `aoti_torch_is_defined` and `aoti_torch_empty_strided_pinned`, neither of which existed here. The first answers whether a tensor holds storage. The second allocates ordinary host memory, because there is no pinned allocator in this runtime and pinning only lets a copy to the device overlap other work. It refuses a device other than CPU, as PyTorch's own does.

**Derived shim spellings.** Some custom operators now reach the fallback-kernel check under the name Inductor derives rather than the name they are registered under. Both spellings are accepted for the Metal ops and the CUDA int4 pack matmul.

**Build and CI repairs.** `setup.py bdist_wheel` is gone, so the macOS source build uses the standard frontend. 2.14 is published for ROCm 7.2, not 7.1. PyTorch now compiles at C++20, which makes CMake scan for modules with scanners the images do not have. And a pip `cmake` in the build environment made the image build silently produce a wheel with no BLAS, so the image's own cmake is used and the result is now asserted rather than assumed.

Two of these were already broken before this branch: the unpinned `katex` install, and the `cmake` interaction.

## Behaviour changes

The re-sync is not cosmetic. `overflows()` answers differently for float to integer casts, in both directions. Filling an int8 tensor with 127.5 was refused and now gives 127. A value at 2^63 cast to int64 was accepted and wrapped, and is now refused.

The portable kernels reach this through `check_overflow_cast`, so `full`, `full_like`, `fill`, `scalar_tensor`, `hardtanh`, `leaky_relu`, `scatter` and `constant_pad_nd` inherit it. The header is a faithful copy of upstream, so this is not ours to undo, but it should be visible rather than buried in a header diff. Neither direction had a test; both do now.

Two smaller ones come with it. Building a delegate sorts its placeholders, and lifted tensor
constants were sorted as if they were user inputs, which left a constant after one and made
any later constant insert impossible. They now sort with the parameters and buffers.

And dropout was missing from the list of operators that take their input's observer. It is an
identity once a model is not training, which is how a quantized model is deployed, so measuring
it separately gave the operator after it a different scale. XNNPACK then refused a reshape whose
two sides disagreed. Both spellings are listed, as several other operators already are.

## Test plan

Each commit carries its own. Covering the whole change:

`compare_dirs.sh`, the header check CI runs, passes against a real `release/2.14` checkout and fails without the re-sync. All eight vendored headers are byte-identical to upstream; the one build-file edit follows a file upstream moved.

The behaviour change was measured by compiling `overflows()` from both branches, not read off the diff. Both new cases fail on the old header, so neither is vacuous.

The regenerated import library was checked member by member: 45 advertised names against 42 before, none lost, and its archive metadata is zeroed so the file is reproducible.

Not covered locally: the image build, the CUDA, ROCm, Qualcomm and Metal jobs, the export suites, the Buck build, and anything needing Windows. CI has all of it.

## Known outstanding

Three things this change does not repair. None of them is new here, and each needs its own
change rather than being folded into a version bump.

No job links the checked-in Windows link stub with the Microsoft linker. The cross build uses
a GNU linker, the same family that produced the archive, so it cannot answer whether the
Microsoft one accepts it. That needs a Windows lowering job, which does not exist yet. There
is now a note beside the file describing how it is produced.

The two Cortex-M model tests expect three quantize pairs where they used to expect one. Two of
those three are round trips: the value is converted back to float and immediately converted
again at the same scale, measured as 0.004997437 and 0.0000032501507 on both sides. So a pass
that used to absorb them no longer matches. The pad operator those pairs sit around is created
by this backend's own passes, after the shared pass that folds such pairs has already run, so
absorbing them means changing the order or extending a pass another backend shares. Correct
output, more work than needed.

Asking for pinned host memory gives ordinary host memory. There is no pinned allocator in this
runtime, and adding one needs a matching release path, so the entry point allocates ordinary
memory and says so. A caller that wants a copy to the device to overlap other work will not get
the overlap.

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Review by digantdesai — APPROVED** [review]

**Comment by digantdesai** [comment]
@claude review this PR.

File: backends/cortex_m/test/models/test_ds_cnn.py [hunk 1]
```diff
@@ -15,7 +15,6 @@
     "executorch_exir_dialects_edge__ops_aten_linear_default": 1,
     "executorch_exir_dialects_edge__ops_aten_relu_default": 9,
     "executorch_exir_dialects_edge__ops_aten_view_copy_default": 1,
-    "executorch_exir_dialects_edge__ops_dim_order_ops__clone_dim_order_default": 2,
```
Comment (rascani) [inline backends/cortex_m/test/models/test_ds_cnn.py:18]: Why did these get removed? Not complaining, just curious. :)

Comment (shoumikhin) [reply backends/cortex_m/test/models/test_ds_cnn.py:18]: Good question, and worth writing down.

On PyTorch 2.14 these graphs no longer contain a dim order clone at all, so the
count cannot match any number. The check is an exact-count assertion, so the entry
has to go rather than change value.

Two things I checked before deleting them, since a vanished op can also mean
something upstream silently stopped working:

The same tests pass on the base branch with these entries present, so this is the
new PyTorch changing the graph rather than an existing problem.

It is not caused by either of the recent dim order passes on main, the one that
removes non-contiguous dim order and the one that replaces input dim order clones
with permutations. Both are already in the base commit, and these tests are green
there.

If the clone is expected to survive for this backend, then this is worth a closer
look rather than a test update, and I am happy to dig further. From what I can see
the clone was only there to make the input contiguous, and 2.14 gets to the same
graph without it, so dropping the expectation looks right.

**Comment by shoumikhin** [comment]
@claude review
